### PR TITLE
Add StableDiffusion random walks tutorial

### PR DESCRIPTION
**Pull request by ianstenbit**
This currently covers three types of walks:
1. Interpolating between two prompt encodings
2. A random walk through the encoder's latent manifold
3. A circular walk through the diffusion noise space

Number 2 has sorta meh results in my opinion, but numbers 1 and 3 are very cool. I think 2 is interesting because it demonstrates some interesting behaviors of stable diffusion (namely that there's a lot of space in the encoder's latent manifold that diffuse to garbage images)

### CODE REVIEW

**Review by LukeWood — CHANGES_REQUESTED** [review]
Requested changes but I'll do a pass of my own; no need to take action.

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+
+keras.mixed_precision.set_global_policy("mixed_float16")
+model = keras_cv.models.StableDiffusion(jit_compile=True)
+
+"""
+## Interpolating Between Text Prompts
+
+In stable diffusion, a text prompt is encoded, and that encoding is used to
+guide the diffusion process. The latent manifold of this encoding is vast, and
```
Comment (LukeWood) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:69]: Instead of vast let’s add the actual dimension of the encoding; we can even print it

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+"""
+Now that we've generated some interpolated images, let's take a look at them!
+
+Throughout this tutorial, we're going to export sequences of images as gifs so
+that they can be easily viewed with some temporal context. For sequences of
+images where the first and last images don't match conceptually, we rubber-band
+the gif.
+"""
+
+def export_as_gif(filename, images, frames_per_second=20, rubber_band=False):
```
Comment (LukeWood) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:110]: The fact that I have looked so many places on the internet and never found a to_gif function that is so simple is baffling.

great job!

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+export_as_gif("doggo-and-fruit-5.gif", [Image.fromarray(img) for img in images],
+              frames_per_second=2, rubber_band=True)
+
+"""
+![Dog to Fruit 5](https://imgur.com/a/617Qjxb)
+
+The results may seem surprising. Generally, interpolating between prompts
+produces coherent looking images, and often demonstrate a progressive concept
+shift between the contents of the two prompts.
```
Comment (LukeWood) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:124]: Perhaps we can add: “This is indicative a high quality representation space”

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+Our next experiment will be to go for a random walk around the latent manifold
+starting from a point produced by a particular prompt.
+"""
+
+walk_steps = 160
+batch_size = 16
+step_size = .05
+
+encoding = model.encode_text("A rubber duck swimming in a bowl of cereal")
```
Comment (LukeWood) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:170]: Should we just raw model.encoder.predict?  Or is this too complicated for that (as it relies on the embeddding too)

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        batch_encodings,
+        batch_size=batch_size,
+        num_steps=25,
+        diffusion_noise=noise,
+    )]
+
+export_as_gif("ducky.gif", images, rubber_band=True)
+
+"""
+![Rubber Ducky Goes on a Walk](https://imgur.com/a/hZAlHrz)
```
Comment (LukeWood) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:191]: Let’s slow these gifs down a bit - the latent space is super smooth and it would be cool to see that

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 4 above)
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py:170]: IMO that would get a little ugly since this code would then have to deal with the pos_ids for the encoder as well as tensorification of the string. Maybe in the long run we should update the API to make model.encoder.encode or something visible, but in its current state I think this is the best option

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 3 above)
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py]: Done

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 1 above)
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py]: Done

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 5 above)
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py:195]: I'm re-running now to slow these down (and use a new seed, because I realized that my original gifs were created with a random seed and I want them to be reproducible with this exact code)

**Comment by LukeWood** [comment]
Two things I’d like to add (I’ll implement these sometime today):

- a gif of only walking one latent dimension
- a grid of a 2d latent walk

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 4 above)
Comment (LukeWood) [reply guides/keras_cv/random_walks_with_stable_diffusion.py:174]: Seems reasonable!

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+quality representation space.
+
+To best visualize this, we should do a much more fine-grained interpolation,
+using hundreds of steps. In order to keep batch size small (so that we don't
+OOM our accelerators), this requires manually batching our interpolated
+encodings.
+"""
+
+interpolation_steps = 160
+batch_size = 16
```
Comment (LukeWood) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:136]: update batch size to 3 to run on Colab

**Comment by LukeWood** [comment]
@fchollet PTAL!  @ianstenbit and I have prepared a strong version of the guide.

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+prompt_2 = "A still life DSLR photo of a bowl of fruit"
+interpolation_steps = 5
+
+encoding_1 = tf.squeeze(model.encode_text(prompt_1))
+encoding_2 = tf.squeeze(model.encode_text(prompt_2))
+
+interpolated_encodings = tf.linspace(encoding_1, encoding_2, interpolation_steps)
+
+# Show the size of the latent manifold
+print(encoding_1.shape)
```
Comment (fchollet) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:87]: Please format print statement to include what is being printed

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+Authors: Ian Stenbit, [fchollet](https://twitter.com/fchollet), [lukewood](https://twitter.com/luke_wood_ml)
+Date created: 2022/09/28
+Last modified: 2022/09/28
+Description: Explore the latent manifold of Stable Diffusion
+"""
+
+"""
+## Overview
+
+Generative models learn a low dimensional latent representation of their
```
Comment (fchollet) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:12]: A concept worth introducing here is that of continuity. The latent space is not just low-dimensional, it is continuous, meaning that you can move from any point to any other via a path where each point is on the manifold (i.e. is a plausible image).

This simple fact is profound and has more to do with the organization of our visual reality than with deep learning. The model learns a latent manifold (and there can be many such manifolds), but the fact that such manifolds exist is remarkable on its own. The visual world is an entirely connected space. This is the reason deep learning works at all on image data.

(One thought experiment to evidence the existence of a "natural" manifold of the real world is the handheld camera thought experiment: you can generate a continuous visual path from any scene to any other scene by keeping an handled camera rolling while you move from the first viewpoint to the new one, e.g. start with the bowl of fruit then keep the camera rolling as you move to your backyard where your dog is sleeping.)

**Review by fchollet — COMMENTED** [review]
Awesome content! Should I just do copyediting pass?

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+using hundreds of steps. In order to keep batch size small (so that we don't
+OOM our accelerators), this requires manually batching our interpolated
+encodings.
+"""
+
+interpolation_steps = 160
+batch_size = 3
+
+encoding_step = (encoding_2 - encoding_1) / interpolation_steps
+encoding_step = tf.squeeze(encoding_step)
```
Comment (ianstenbit) [inline guides/keras_cv/random_walks_with_stable_diffusion.py:158]: I've realized that we can probably afford from a memory perspective to just generate the entire encoding batch here instead of doing the ugly intermediate interpolation steps.

Would be:

```
interpolated_encodings = tf.linspace(encoding1, encoding2, interpolation_steps)
encoding_batches = tf.split(interpolated_encodings, interpolation_steps // batch_size)

for index in range(interpolation_steps // batch_size):
    encodings = encoding_batches[index]
    images += [.....
```

I need to focus on benchmark tests rn but I can add this tomorrow afternoon if you don't get to it @LukeWood 


```

**Comment by LukeWood** [comment]
> Awesome content! Should I just do copyediting pass?

Please do!

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 9 above)
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py]: Okay I've updated this and re-run all the experiments to verify it works.

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 6 above)
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py]: Done throughout

File: guides/keras_cv/random_walks_with_stable_diffusion.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
+"""
+Title: Random Walks with Stable Diffusion in KerasCV
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply guides/keras_cv/random_walks_with_stable_diffusion.py]: @ianstenbit any opposition to this?

I'm on board with:

"A Walk through latent space with Stable Diffusion"

**Review by fchollet — COMMENTED** [review]
Looking great! I pushed some copyedits. I think we're ready to add the generated files?

File: guides/keras_cv/random_walks_with_stable_diffusion.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ianstenbit) [reply guides/keras_cv/random_walks_with_stable_diffusion.py]: sg -- this is now the title

**Comment by ianstenbit** [comment]
> Looking great! I pushed some copyedits. I think we're ready to add the generated files?

I think so. I will go ahead and generate on a GCP VM.

**Comment by ianstenbit** [comment]
> > Looking great! I pushed some copyedits. I think we're ready to add the generated files?
> 
> I think so. I will go ahead and generate on a GCP VM.

I've now rendered the guide using the rendering script.

**Review by LukeWood — APPROVED** [review]
LGTM thanks!  Excited to launch this

**Review by fchollet — APPROVED** [review]
👍 

Will post this tomorrow and repost on Monday (since week ends can be slow...)

**Comment by fchollet** [comment]
@ianstenbit can you send me all of the gifs over email?

**Comment by ianstenbit** [comment]
> @ianstenbit can you send me all of the gifs over email?

Yes -- I've sent this to both you and Luke

**Comment by LukeWood** [comment]
Thanks guys!  Super cool
